pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+                    ra_shift, dec_shift = calculate_image_phase(reference_image.copy(), wl_imgs[:, :, ff], maskval=0.0)
+                    # Convert pixel shift to degrees shift
+                    ra_shift *= self._dspat/cosdec
+                    dec_shift *= self._dspat
+                    msgs.info("Spatial shift of cube #{0:d}: RA, DEC (arcsec) = {1:+0.3f} E, {2:+0.3f} N".format(ff+1, ra_shift*3600.0, dec_shift*3600.0))
+                    # Apply the shift
+                    self.all_ra[self.all_idx == ff] += ra_shift
+                    self.all_dec[self.all_idx == ff] += dec_shift
+
+    def compute_weights(self):
+        """
+        Compute the relative weights to apply to pixels that are collected into the voxels of the output DataCubes
+        """
+        # Calculate the relative spectral weights of all pixels
+        if self.numfiles == 1:
+            # No need to calculate weights if there's just one frame
+            self.all_wghts = np.ones_like(self.all_sci)
+        else:
+            # Find the wavelength range where all frames overlap
+            min_wl, max_wl = datacube.get_whitelight_range(np.max(self.mnmx_wv[:, :, 0]),  # The max blue wavelength
+                                                  np.min(self.mnmx_wv[:, :, 1]),  # The min red wavelength
+                                                  self.cubepar['whitelight_range'])  # The user-specified values (if any)
+            # Get the good white light pixels
+            ww, wavediff = datacube.get_whitelight_pixels(self.all_wave, min_wl, max_wl)
+            # Get a suitable WCS
+            image_wcs, voxedge, reference_image = self.create_wcs(self.all_ra, self.all_dec, self.all_wave,
+                                                                  self._dspat, wavediff, collapse=True)
+            # Generate the white light image (note: hard-coding subpixel=1 in both directions, and combining into a single image)
+            wl_full = generate_image_subpixel(image_wcs, self.all_ra, self.all_dec, self.all_wave,
+                                              self.all_sci, self.all_ivar, self.all_wghts,
+                                              self.all_spatpos, self.all_specpos, self.all_spatid,
+                                              self.all_tilts, self.all_slits, self.all_align, voxedge, all_idx=self.all_idx,
+                                              spec_subpixel=1, spat_subpixel=1, combine=True)
+            # Compute the weights
+            self.all_wghts = datacube.compute_weights(self.all_ra, self.all_dec, self.all_wave, self.all_sci, self.all_ivar, self.all_idx, wl_full[:, :, 0],
+                                                      self._dspat, self._dwv, relative_weights=self.cubepar['relative_weights'])
+
+    def coadd(self):
+        """
+        This is the main routine called to convert PypeIt spec2d files into PypeIt DataCube objects. It is specific
+        to the SlicerIFU data.
```

Comment:
Should this be a core method.

Reply:
I think this should stay here, but some of the functions called in this section can go into core. I've really just used this to break up individual parts of the code

Reply:
I'm fine with this method remaining in the class, but am pushing back on self.compute_weights which I think should be in core. Please also see my comment about return values to make the control flow more clear.